pennylane/templates/subroutines/fable.py

```diff
@@ -0,0 +1,177 @@
+    one can add zeros to reach the correct dimension.
+    """
+    num_wires = AnyWires
+    """int: Number of wires that the operator acts on."""
+
+    num_params = 1
+    """int: Number of trainable parameters that the operator depends on."""
+
+    grad_method = None
+    """Gradient computation method."""
+
+    def __init__(self, A, tol=0, id=None):
+        M, N = A.shape
+        if M != N:
+            raise ValueError(
+                f"The input matrix should be of shape NxN, got {A.shape}. Try padding zeros for the correct dimension"
+            )
+
+        alpha = np.linalg.norm(np.ravel(A), np.inf)
+        if alpha > 1:
+            raise ValueError(
+                "The subnormalization factor should be lower than 1. Ensure that the values of the input matrix are within [-1, 1]"
+            )
+
+        if np.any(np.iscomplex(A)):
+            warnings.warn(
+                "Support for imaginary values has not been implemented yet. Proceed with caution."
+            )
+
+        self._hyperparameters = {"tol": tol}
+
+        ancilla = ["ancilla"]
+        s = int(np.log2(np.array(A).shape[0]))
+        wires_i = [f"i{index}" for index in range(s)]
+        wires_j = [f"j{index}" for index in range(s)]
+
+        all_wires = Wires(ancilla) + Wires(wires_i) + Wires(wires_j)
+        super().__init__(A, wires=all_wires, id=id)
+
+    @staticmethod
+    def compute_decomposition(A, wires, tol=0):  # pylint:disable=arguments-differ
+        r"""Sequence of gates that represents the efficient circuit produced by the FABLE technique
+
+        Args:
+            A (tensor_like): an :math:`(N \times N)` matrix to be encoded
+            wires (Any or Iterable[Any]): wires that the operator acts on
+            tol (float): tolerance
+
+        Returns:
+            list[.Operator]: list of gates for efficient circuit
+        """
+        op_list = []
+        alphas = np.arccos(A).flatten()
+        thetas = compute_theta(alphas)
+
+        ancilla = [wires[0]]
+        wires_i = wires[1 : 1 + len(wires) // 2]
+        wires_j = wires[1 + len(wires) // 2 : len(wires)]
+
+        code = gray_code(2 * np.sqrt(len(A)))
+        n_selections = len(code)
+
+        control_wires = [
+            int(np.log2(int(code[i], 2) ^ int(code[(i + 1) % n_selections], 2)))
+            for i in range(n_selections)
+        ]
+
+        wire_map = dict(enumerate(wires_j + wires_i))
+
+        for w in wires_i:
+            op_list.append(qml.Hadamard(w))
+
+        nots = []
+        for theta, control_index in zip(thetas, control_wires):
+            if abs(2 * theta) > tol:
+                for c_wire in nots:
+                    op_list.append(qml.CNOT(wires=[c_wire] + ancilla))
+                op_list.append(qml.RY(2 * theta, wires=ancilla))
+                nots = []
+            if (cw := wire_map[control_index]) in nots:
+                del nots[nots.index(cw)]
+            else:
+                nots.append(wire_map[control_index])
```

Comment:
Could you explain what is happening in this code block? Why are we iteratively populating the nots list and deleting elements from it?

Reply:
The goal is to only add RY gates that are above a certain tolerance and to remove CNOT gates that cancel each other out. 

For the RY part, we should only append if the theta is greater than the tolerance. For the CNOT part however we need to delete the pairs that cancel each other out up until the next RY gate (since the CNOT gates no longer cancel each other out). The nots list keeps track of the non-paired CNOTs we have before our next RY above the tolerance. We delete the paired (cancellable) CNOTs from our nots list.

For example, suppose tol = 0.01, and our thetas are 0.3, 0, 0, 0.2 and our CNOTS are (assuming ancilla is the root) [j0, j1, j0, i0]. Our nots list at the end of each iteration looks like this:
Iteration 1. Since 0.3 > 0.01, we append RY to the op_list. We append j0 to nots. (nots = [j0])
Iteration 2. Since 0 < 0.01, we do not yet append the CNOTs from our nots list. We check to see if this CNOT gate can be cancelled, if so, we delete. In this case j1 is not in nots, so instead we append j1 to nots. (nots = [j0, j1])
Iteration 3. Since 0 < 0.01, we do not yet append the CNOTs from our nots list. We check to see if this CNOT gate can be cancelled. We find that j0 is in the nots list, we can del j0 from there. (nots = [j1])
Iteration 4. Since 0.2 < 0.01, we append the CNOT gate (ancilla -> j1) to the op_list. We also reset the nots array to []. We then append i0 to the nots list. nots = [i0].

Although since nots is a list, the search and deletion through the list are both O(N) operations. This can be improved by using a dictionary, where search and delete operations are O(1).

Reply:
Awesome, that sounds about right to me. As you correctly pointed out, for each RY gate (or iteration) you are potentially doing an O(n) search and delete for each Cnot gate. So the total run time is something like O(n*d) where d is the number of angles. You can see how this could be prohibitively slow when trying to execute large encodings of hundreds of qubits. Could you please implement the strategy you suggested to reduce the run time? 

As a follow up question (no need to code this), this code is still very "serial" in the sense that we need to iterate through each operation one after the next and append them to the final list of ops to be executed. Could you describe a way to parallelize this logic? Maybe something along the lines of a divide and conquer?

Reply:
Yes I've implemented it using a dictionary instead now. 

For the follow up, my first instinct is that whenever you have a group of thetas below the tolerance, you can process this as a batch in parallel. We can illustrate this idea with the previous example.

In the previous example, (0.3, 0.0, 0.0, 0.2), we can divide the angles into two batches. The first being (0.3, 0.0, 0.0) and the second being (0.2). We can execute these in parallel so long as we ensure that the ordering of the batches remains the same. So process one will work on (0.3, 0.0, 0.0) and append to the op_list, whereas process two will work on (0.2) and, if needed, wait until process one is finished and append to the op_list. The ordering of the batches cannot be ignored because all the CNOTs use the ancilla qubit as the control bit.

However, this implementation requires a pre-processing step of finding angles over the threshold to divide it into batches. I have thought about this for a bit but I dont think it is possible to do this without any pre-processing whatsoever.